### DISCUSSION TITLE
How to handle datetime-local inputs

### DISCUSSION THREAD

**Discussion by lifeiscontent** [edmundhung/conform#738] [Q&A]
@edmundhung curious to get your take on this.

I'm currently struggling with understanding how to handle datetime-local values

with datetime-local it consumes a ISO8601 date without the `Z` at the end of it

based off my understanding, getInputProps may not be the right place for it because we're transforming values and we can't be 100% sure that people would be passing a full ISO8601 datetime, in my application that's what we use, but in order to allow anyone with whatever they might need it should be general

my question to you is how would you handle something like this and at which boundary would you have some logic like this in the pipeline of input/outputs

```tsx
import type { FieldMetadata } from "@conform-to/react";
import { getFormProps, getInputProps, useForm } from "@conform-to/react";
import { getZodConstraint, parseWithZod } from "@conform-to/zod";
import type { ActionFunctionArgs } from "@remix-run/node";
import { json } from "@remix-run/node";
import { Form } from "@remix-run/react";
import { z } from "zod";

function fieldForDatetimeLocal(field: FieldMetadata) {
  return new Proxy(field, {
    get(target, prop, receiver) {
      if (prop === "initialValue") {
        const initialValue = Reflect.get(target, prop, receiver);
        if (typeof initialValue !== "string") {
          return initialValue;
        }
        const utcOrLocal = new Date(initialValue);
        // initialValue could potentially be a stringified number, in that case this code will be wrong
        const local = initialValue.endsWith("Z")
          ? new Date(
              utcOrLocal.getTime() - utcOrLocal.getTimezoneOffset() * 60000,
            )
          : new Date(initialValue);

        if (local.toString() === "Invalid Date") {
          return initialValue;
        }
        return local.toISOString().slice(0, -1);
      }
      return Reflect.get(target, prop, receiver);
    },
  });
}

const schema = z.object({
  date: z.preprocess((x) => {
    const date = typeof x === "number" ? new Date(x) : new Date(String(x));
    if (date.toString() === "Invalid Date") {
      return x;
    }
    return date.toISOString();
  }, z.string().datetime().optional().default(z.NEVER)),
});

export async function action({ request }: ActionFunctionArgs) {
  const formData = await request.formData();
  const submission = parseWithZod(formData, { schema });

  if (submission.status !== "success") {
    return json(submission.reply());
  }

  console.log(submission.value);

  return json(submission.reply());
}

export default function Example() {
  const [form, fields] = useForm({
    constraint: getZodConstraint(schema),
    defaultValue: {
      date: new Date(0).toISOString(),
    },
    onValidate(context) {
      return parseWithZod(context.formData, { schema });
    },
    shouldRevalidate: "onInput",
    shouldValidate: "onBlur",
  });

  return (
    <Form {...getFormProps(form)} method="POST">
      <input
        {...getInputProps(fieldForDatetimeLocal(fields.date), {
          type: "datetime-local",
        })}
      />
      <button type="submit">Submit</button>
    </Form>
  );
}
```

**Comment by edmundhung** [comment]
I can see why you struggle. I think `getInputProps` is the right place to do this, but it is really hard to differentiate whether the value is properly serialized or not at the moment. What if Conform stop serializing the value for you? That way it could be easier to identify serialized and non-serialized value (e.g. `string` vs `Date`). But then how should we do if the data source is a string as well? 🤔  

A less ideal but simple approach is to make yourself a simple util like `serializeDateTime()` and make sure the value is serialized before passing to Conform:

```tsx
export function serializeDateTime(date: Date): string {
  // ...
}

export default function Example() {
  const [form, fields] = useForm({
    constraint: getZodConstraint(schema),
    defaultValue: {
       // serialize the value before passing to conform
      date: serializeDateTime(new Date(0)),
    },
    onValidate(context) {
      return parseWithZod(context.formData, { schema });
    },
    shouldRevalidate: "onInput",
    shouldValidate: "onBlur",
  });

  return (
    <Form {...getFormProps(form)} method="POST">
      <input
        {...getInputProps(fields.date, {
          type: "datetime-local",
        })}
      />
      <button
        {...form.update.getButtonProps({
          name: fields.date.name,
          // serialize the value before passing to conform..
          value: serializeDateTime(new Date(0))
        })}
      >
        Update date
      </button>
    </Form>
  );
}
```

**Reply by lifeiscontent** [reply]
@edmundhung so the value that comes back, at least from my API is a ISO8601 string, not a date, as its data from a JSON API.

I like that Conform serializes the data for me, it does it in ways that I'd expect.

one idea is to take some inspiration from Formik they were in the process of creating a parse/format API which we could potentially do with `getInputProps` https://github.com/jaredpalmer/formik/pull/3152

the one issue I am currently seeing is if we did this, the `dirty` check for fields might get out of sync.

if you have ideas on how to solve the dirty issue, I think the parse/format helpers for `get{Input|etc}Props` could be a nice solution

**Reply by lifeiscontent** [reply]
also, in the scenario that the `defaultValue` is a large JSON object that comes back from an API, it would be a pain to pass in a formatter for each field

**Comment by lifeiscontent** [comment]
@edmundhung on second thought, the value parsing conform does might be better if I could pull it out, I'm gonna send you a schema on discord for review. but basically because of how I've defined a schema, boolean values are improperly serialized.
